pkg/crd/crd.go

```diff
@@ -0,0 +1,136 @@
+// Package crd helps with the installation of CustomResourceDefinitions.
+package crd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clientv1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	waitDuration = 1 * time.Minute
+	waitInterval = 500 * time.Millisecond
+)
+
+// BatchCreateCRDs will create or update the list of crds if they do not exist.
+func BatchCreateCRDs(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface, crds []*apiextv1.CustomResourceDefinition) error {
+	existingCRDs, err := getExistingCRD(ctx, crdClient)
+	if err != nil {
+		// we do not need to return this error because we can still attempt to create the CRDs if the list fails
+		logrus.Warnf("unable to check for existing CRDs: %s", err.Error())
+	}
+
+	// ensure each CRD exist
+	for _, crd := range crds {
+		err = ensureCRD(ctx, crd, crdClient, existingCRDs)
+		if err != nil {
+			return err
+		}
+	}
+
+	existingCRDs, err = getExistingCRD(ctx, crdClient)
+	if err != nil {
+		return fmt.Errorf("unable to check for existing CRDs: %w", err)
+	}
+	// ensure each CRD is ready
+	for _, crd := range crds {
+		if existingCRD, ok := existingCRDs[crd.Name]; ok && crdIsReady(existingCRD) {
+			continue
+		}
+		// wait for the CRD to be ready
+		err := waitCRD(ctx, crd.Name, crdClient)
+		if err != nil {
+			return fmt.Errorf("failed waiting on CRD '%s': %w", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
```

Comment:
Since this gets all existing CRDs, it should be plural
```suggestion
func getExistingCRDs(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
```